core/src/main/scala/org/http4s/headers/Access-Control-Allow-Methods.scala

```diff
@@ -7,4 +7,26 @@
 package org.http4s
 package headers
 
-object `Access-Control-Allow-Methods` extends HeaderKey.Default
+import org.http4s.util.Writer
+import org.http4s.parser.HttpHeaderParser
+
+object `Access-Control-Allow-Methods`
+    extends HeaderKey.Internal[`Access-Control-Allow-Methods`]
+    with HeaderKey.Singleton {
+
+  def apply(methods: Method*): `Access-Control-Allow-Methods` =
+    `Access-Control-Allow-Methods`(methods.toSet)
+
+  override def parse(s: String): ParseResult[`Access-Control-Allow-Methods`] =
+    HttpHeaderParser.ACCESS_CONTROL_ALLOW_METHODS(s)
+}
+
+final case class `Access-Control-Allow-Methods`(methods: Set[Method]) extends Header.Parsed {
```

Comment:
It looks like you're using `Set.empty` as the wildcard value.  I wonder about being more explicit in the type: `Either[Wildcard, NonEmptySet[Method]]`.  It's more verbose, but also more meaningful.  Or could even create a sum type other than `Either`, to avoid arbitrary `Left` and `Right`.

Reply:
I'm also not sure how I feel about the `Set`.  Nothing says that they have to be unique (even if it's silly for them not to be), and we lose the ability to render them exactly the way we got them.  Maybe a `List` (or `NonEmptyList`, pending the above) instead of a `Set`?  I could go either way.

Reply:
@rossabaker thank you for the reviews and comment. I do agree going with an explicit sum type would be better and more readable. With that in mind I was hoping to create something like this

```scala
sealed trait WildCardOrMethods

final case class Wildcard() extends WildCardOrMethods
final case class Methods(methods: NonEmptyList[Method]) extends WildCardOrMethods
```

And the parser would then check if there is even one wild card in the list, if there is `Wilcard` is passed to `Access-Control-Allow-Methods` case class. If not, the parser passes the parsed methods to the case class.

Am I thinking in the right direction?